service/worker/service.go

```diff
@@ -185,15 +185,21 @@ func NewConfig(params *resource.Params) *Config {
 	advancedVisWritingMode := dc.GetStringProperty(
 		dynamicconfig.AdvancedVisibilityWritingMode,
 	)
+	// only start indexer when advanced visibility writing mode is set to on and ES visibility store is configured
+	// when it is using pinot and not in migration mode, indexer should not be started
 	if common.IsAdvancedVisibilityWritingEnabled(advancedVisWritingMode(), params.PersistenceConfig.IsAdvancedVisibilityConfigExist()) {
-		config.IndexerCfg = &indexer.Config{
-			IndexerConcurrency:             dc.GetIntProperty(dynamicconfig.WorkerIndexerConcurrency),
-			ESProcessorNumOfWorkers:        dc.GetIntProperty(dynamicconfig.WorkerESProcessorNumOfWorkers),
-			ESProcessorBulkActions:         dc.GetIntProperty(dynamicconfig.WorkerESProcessorBulkActions),
-			ESProcessorBulkSize:            dc.GetIntProperty(dynamicconfig.WorkerESProcessorBulkSize),
-			ESProcessorFlushInterval:       dc.GetDurationProperty(dynamicconfig.WorkerESProcessorFlushInterval),
-			ValidSearchAttributes:          dc.GetMapProperty(dynamicconfig.ValidSearchAttributes),
-			EnableQueryAttributeValidation: dc.GetBoolProperty(dynamicconfig.EnableQueryAttributeValidation),
+		if params.PersistenceConfig.AdvancedVisibilityStore == common.PinotVisibilityStoreName && params.PinotConfig != nil && !params.PinotConfig.Migration.Enabled {
```

Comment:
Maybe this should be pulled out into a simple helper? 

```
if shouldStartIndexer(params) {
    config.IndexerCfg = &indexer.Config{
        IndexerConcurrency:             dc.GetIntProperty(dynamicconfig.WorkerIndexerConcurrency),
        ESProcessorNumOfWorkers:        dc.GetIntProperty(dynamicconfig.WorkerESProcessorNumOfWorkers),
        ESProcessorBulkActions:         dc.GetIntProperty(dynamicconfig.WorkerESProcessorBulkActions),
        ESProcessorBulkSize:            dc.GetIntProperty(dynamicconfig.WorkerESProcessorBulkSize),
        ESProcessorFlushInterval:       dc.GetDurationProperty(dynamicconfig.WorkerESProcessorFlushInterval),
        ValidSearchAttributes:          dc.GetMapProperty(dynamicconfig.ValidSearchAttributes),
        EnableQueryAttributeValidation: dc.GetBoolProperty(dynamicconfig.EnableQueryAttributeValidation),
    }
}
```

Reply:
Fixed!